dra-evolution/pkg/api/capacity_types.go

```diff
@@ -3,168 +3,110 @@ package api
+type ResourcePool struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
 	// +optional
-	NodeName *string `json:"nodeName,omitempty"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
 
-	// Driver is the name of the DeviceDriver that created this object and
-	// owns the data in it.
-	// +required
-	Driver string `json:"driver,omitempty"`
+	Spec ResourcePoolSpec `json:"spec"`
 
-	// Attributes contains device attributes that are common to all devices
-	// in the pool.
-	// +optional
-	// +listType=atomic
-	Attributes []Attribute `json:"attributes,omitempty"`
-
-	// SharedResources are pooled resources that are shared by all devices
-	// in the pool. This is typically used when representing a
-	// partitionable device, and need not be populated otherwise.
+	// Stretch goal for 1.31: define status
 	//
-	// +optional
-	// +listType=atomic
-	SharedResources []ResourceCapacity `json:"sharedResources,omitempty"`
+	// To discuss:
+	// - Who writes that status?
+	//   After https://github.com/kubernetes/kubernetes/pull/125163 (implements
+	//   https://github.com/kubernetes/enhancements/pull/4667), kubelet is not
+	//   involved with publishing ResourcePools, they come directly from the driver.
+	// - What information should be in it?
+	// - Does it need to be a sub-resource? This would make it harder
+	//   for a driver to publish a new device (spec) and its corresponding
+	//   health information (status).
+}
 
-	// Devices contains the individual devices in the pool.
+type ResourcePoolSpec struct {
+	// NodeName identifies the node which provides the devices. All devices
+	// are local to that node.
 	//
-	// +required
-	// +listType=atomic
-	Devices []Device `json:"devices,omitempty"`
-}
+	// This is currently required, but this might get relaxed in the future.
+	NodeName string `json:"nodeName"`
 
-// DevicePoolStatus contains the state of the pool as last reported by the
-// driver. Note that this will not include the allocations that have been made
-// by the scheduler but not yet seen by the driver. Thus, it is NOT sufficient
-// to make future scheduling decisions.
-type DevicePoolStatus struct {
-	DeviceStatuses []AllocatedDevice `json:"deviceStatuses,omitempty"`
-}
+	// POTENTIAL FUTURE EXTENSION: NodeSelector *v1.NodeSelector
 
-// AllocatedDevice represents a device that has been allocated from the pool.
-type AllocatedDevice struct {
-	Name string
-	// Conditions contains the latest observation of the device's state.
-	Conditions []metav1.Condition `json:"conditions"`
+	// DriverName identifies the DRA driver providing the capacity information.
+	// A field selector can be used to list only ResourcePool
+	// objects with a certain driver name.
+	//
+	// Must be a DNS subdomain and should end with a DNS domain owned by the
+	// vendor of the driver.
+	DriverName string `json:"driverName" protobuf:"bytes,3,name=driverName"`
+
+	// Devices lists all available devices in this pool.
```

Comment:
We need to spec a max number of records.  Something like 128 or 256 should be good?  Even 1024 should be OK, but now we are at risk of hitting etcd limits.

This may be an argument for bringing in a pool-level attributes list (common to all devices).  That also takes some of the air out of the "gizmo" discussion, unless gizmos also have attributes?

Reply:
I know we should limit it, but how? If we are conservative at all levels (128 devices, 32 attributes each), then we will prevent creating pools which would fit the etcd limits (e.g. 1 device, 128 attributes).

I just randomly picked 128 + 32. I'm not even sure whether it would be sufficiently small. The maximum request body size in the apiserver is 3MB, twice of what is recommended (required?) for etcd requests (https://github.com/kubernetes/apiserver/blob/8d18eec7c050338aac4d49e470f3ea0b946f4726/pkg/server/config.go#L433-L442).

I', wondering whether it would be okay to not limit array sizes and just say "total size must not exceed 1MB". That leaves 0.5MB for status, which might be written by others. But I don't know how to implement that and it probably wouldn't be okay also because "excessive" array sizes could also be bad in other ways.

So, back of the envelope calculation: one attribute has a name with DNS label (63 bytes) + identifier (length to be determined) - let's say 80 bytes overall, including value. 256 devices with 64 attributes ~= 1.25MB. 200 + 50 ~= 0.76MB. That leaves room for device name (not included yet) and status.

Reply:
Pool-level attributes don't help with allowing higher limits because the worst case is still the same as without them. I see that as a future enhancement to minimize the average size, and I am not sure whether it's worth it. Need more data...

I'm okay with 128 + 32 as a start. @klueska ?